File: my-app/backend/routes/threads.js
const express = require('express');
const router = express.Router();
const { body, param, validationResult } = require('express-validator');
const Message = require('../models/Message');
const Reply = require('../models/Reply');

/**
 * GET /api/threads/:threadToken
 * Retrieve a thread (original message + all replies) by token.
 * Marks the message as read on first access (self-destruct trigger).
 */
router.get(
  '/:threadToken',
  [
    param('threadToken')
      .isUUID()
      .withMessage('Invalid thread token format'),
  ],
  async (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(422).json({ errors: errors.array() });
    }

    try {
      const { threadToken } = req.params;

      const thread = await Message.findOne({ threadToken }).populate('replies');

      if (!thread) {
        return res.status(404).json({ error: 'Thread not found. Check your token.' });
      }

      // Mark as read on first access
      if (!thread.isRead) {
        thread.isRead = true;
        thread.readAt = new Date();
        await thread.save();
      }

      return res.json({
        success: true,
        thread: {
          threadToken: thread.threadToken,
          topic: thread.isDeleted ? '[deleted]' : thread.topic,
          message: thread.isDeleted ? '[deleted]' : thread.message,
          isDeleted: thread.isDeleted,
          isRead: thread.isRead,
          readAt: thread.readAt,
          createdAt: thread.createdAt,
          replies: thread.replies,
        },
      });
    } catch (err) {
      console.error('Error fetching thread:', err);
      return res.status(500).json({ error: 'Failed to retrieve thread.' });
    }
  }
);

/**
 * POST /api/threads/:threadToken/reply
 * Add a follow-up message to an existing thread.
 */
router.post(
  '/:threadToken/reply',
  [
    param('threadToken')
      .isUUID()
      .withMessage('Invalid thread token format'),
    body('topic')
      .trim()
      .notEmpty()
      .withMessage('Topic is required')
      .isLength({ max: 200 })
      .withMessage('Topic must be 200 characters or fewer'),
    body('message')
      .trim()
      .notEmpty()
      .withMessage('Message is required')
      .isLength({ max: 5000 })
      .withMessage('Message must be 5000 characters or fewer'),
  ],
  async (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(422).json({ errors: errors.array() });
    }

    try {
      const { threadToken } = req.params;
      const { topic, message } = req.body;

      const thread = await Message.findOne({ threadToken });

      if (!thread) {
        return res.status(404).json({ error: 'Thread not found. Check your token.' });
      }

      if (thread.isDeleted) {
        return res.status(410).json({ error: 'This thread has been deleted.' });
      }

      // Create the reply
      const reply = await Reply.create({ threadToken, topic, message });

      // Link reply to thread
      thread.replies.push(reply._id);
      await thread.save();

      return res.status(201).json({
        success: true,
        message: 'Your follow-up has been sent securely.',
        reply: {
          topic: reply.topic,
          createdAt: reply.createdAt,
        },
      });
    } catch (err) {
      console.error('Error adding reply:', err);
      return res.status(500).json({ error: 'Failed to send follow-up. Please try again.' });
    }
  }
);

module.exports = router;

File: my-app/backend-go/internal/app/threads.go
package app

import (
	"database/sql"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
)

func threadsRoutes(store *Store) http.Handler {
	db := store.DB()
	r := chi.NewRouter()

	r.Get("/{threadToken}", func(w http.ResponseWriter, r *http.Request) {
		threadToken := chi.URLParam(r, "threadToken")
		if strings.TrimSpace(threadToken) == "" {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Invalid thread token format"})
			return
		}
		var thread Message
		var reportedStaff sql.NullString
		var readAt sql.NullTime
		var isRead, isDeleted int
		err := db.QueryRowContext(r.Context(),
			"SELECT id, topic, message, reportedStaff, threadToken, isRead, isDeleted, readAt, createdAt FROM messages WHERE threadToken = ?",
			threadToken).Scan(&thread.ID, &thread.Topic, &thread.Message, &reportedStaff, &thread.ThreadToken, &isRead, &isDeleted, &readAt, &thread.CreatedAt)
		if err != nil {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Thread not found. Check your token."})
			return
		}
		if reportedStaff.Valid {
			thread.ReportedStaff = &reportedStaff.String
		}
		thread.IsRead = isRead != 0
		thread.IsDeleted = isDeleted != 0
		if readAt.Valid {
			t := readAt.Time.UTC()
			thread.ReadAt = &t
		}
		thread.CreatedAt = thread.CreatedAt.UTC()

		if !thread.IsRead {
			_, _ = db.ExecContext(r.Context(), "UPDATE messages SET isRead = 1, readAt = ? WHERE id = ?", time.Now().UTC(), thread.ID)
			thread.IsRead = true
		}

		var replies []Reply
		rows, err := db.QueryContext(r.Context(), "SELECT id, threadToken, topic, message, isDeleted, createdAt, updatedAt FROM replies WHERE threadToken = ?", threadToken)
		if err == nil {
			defer rows.Close()
			for rows.Next() {
				var rp Reply
				var createdAt, updatedAt time.Time
				var isDeletedFlag int
				if err := rows.Scan(&rp.ID, &rp.ThreadToken, &rp.Topic, &rp.Message, &isDeletedFlag, &createdAt, &updatedAt); err != nil {
					continue
				}
				rp.IsDeleted = isDeletedFlag != 0
				rp.CreatedAt = createdAt.UTC()
				if !updatedAt.IsZero() {
					rp.UpdatedAt = updatedAt.UTC()
				}
				replies = append(replies, rp)
			}
			_ = rows.Err()
		}

		writeJSON(w, http.StatusOK, map[string]any{"success": true, "thread": map[string]any{"threadToken": thread.ThreadToken, "topic": thread.Topic, "message": thread.Message, "isDeleted": thread.IsDeleted, "isRead": thread.IsRead, "readAt": thread.ReadAt, "createdAt": thread.CreatedAt, "replies": replies}})
	})

	r.Post("/{threadToken}/reply", func(w http.ResponseWriter, r *http.Request) {
		threadToken := chi.URLParam(r, "threadToken")
		if strings.TrimSpace(threadToken) == "" {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Invalid thread token format"})
			return
		}
		var payload struct {
			Topic   string `json:"topic"`
			Message string `json:"message"`
		}
		if err := decodePayload(r, &payload); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Invalid JSON payload"})
			return
		}
		if strings.TrimSpace(payload.Topic) == "" || strings.TrimSpace(payload.Message) == "" {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Topic and message are required"})
			return
		}

		var threadID int64
		var isDeleted int
		err := db.QueryRowContext(r.Context(), "SELECT id, isDeleted FROM messages WHERE threadToken = ?", threadToken).Scan(&threadID, &isDeleted)
		if err != nil {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Thread not found. Check your token."})
			return
		}
		if isDeleted != 0 {
			writeJSON(w, http.StatusGone, map[string]any{"error": "This thread has been deleted."})
			return
		}
		now := time.Now().UTC()
		reply := Reply{ThreadToken: threadToken, Topic: strings.TrimSpace(payload.Topic), Message: strings.TrimSpace(payload.Message), CreatedAt: now}
		_, err = db.ExecContext(r.Context(),
			"INSERT INTO replies (threadToken, topic, message, isDeleted, createdAt, updatedAt) VALUES (?, ?, ?, 0, ?, ?)",
			threadToken, reply.Topic, reply.Message, now, now)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to send follow-up. Please try again."})
			return
		}
		writeJSON(w, http.StatusCreated, map[string]any{"success": true, "message": "Your follow-up has been sent securely.", "reply": map[string]any{"topic": reply.Topic, "createdAt": reply.CreatedAt}})
	})

	return r
}
